### PR TITLE
#1635 Add additional cap exceeded status

### PR DESCRIPTION
**Pull request by drakkein**

### CODE REVIEW

File: app/modules/public-etos/sagas.ts [hunk 1]
```diff
@@ -230,6 +232,30 @@ function* downloadTemplateByType(_: TGlobalDependencies, action: TAction): any {
...
+  const eto: TPublicEtoData = action.payload;
+  const etoContract: ETOCommitment = yield contractsService.getETOCommitmentContract(eto.etoId);
+
+  if (!etoContract) return;
+
+  const from = selectEthereumAddressWithChecksum(state);
+  // TODO: check whether typechain but still is not fixed
+
+  // sorry no typechain, typechain has a bug with boolean casting
+  const calculation = yield promisify(etoContract.rawWeb3Contract.calculateContribution, [
```
Comment (rudolfix) [inline app/modules/public-etos/sagas.ts]: I think we have contributions somewhere already in state? we can reuse them. Also I think it's being used to check if user can participate in whitelisting. So maybe there is a way to reduce number of calls? We can also use state for the investment progress (how many tokens were sold vs maximum) to optimize it. Please ping me on slack so we can clean it up.

Comment (desfero) [reply app/modules/public-etos/sagas.ts]: I see it's almost identical to investor-ticket. We can extract same piecie of code to separate generator and then call both of them with `neuCall`.

```
function* checkMaxCapExceeded({ contractsService }: TGlobalDependencies, action: TAction): any {
  if (action.type !== "PUBLIC_ETOS_CALCULATE_ETO_MAX_CAP_EXCEEDED") return;
  
  const contribution: SOMETYPE = yield neuCall(calculateContribution, from, eto.minTicketEur);

  yield put(
    actions.publicEtos.setMaxCap({ previewCode: eto.previewCode, capExceeded: calculation.pop() }),
  );
}
```

File: app/modules/public-etos/reducer.ts [hunk 2]
```diff
@@ -59,6 +61,14 @@ export const publicEtosReducer: AppReducer<IPublicEtoState> = (
           [action.payload.previewCode]: action.payload.data,
         },
       };
+    case "PUBLIC_ETOS_SET_ETO_MAX_CAP_EXCEEDED":
+      return {
+        ...state,
```
Comment (desfero) [inline app/modules/public-etos/reducer.ts]: @Moejoe90 @drakkein we need to think how to simplify this :D maybe some small library to wrap this into something, i don't know lenses/immutablejs/whatever :D

File: app/styles/variables.scss [hunk 3]
```diff
@@ -26,6 +26,8 @@ $blue-light: #78d7dc;
 $green-light: #89eb82;
 $green2: #bbefe4;
 $green3: #41b5aa;
+$green4: #99e7b0;
+$green5: #2ddd73;
```
Comment (desfero) [inline app/styles/variables.scss]: lol, too many greens, need to talk with designers as we should keep pattern library small

File: app/components/shared/PercentageIndicatorBar.tsx [hunk 4]
```diff
@@ -82,6 +89,16 @@ export const PercentageIndicatorBar: React.SFC<IProps & CommonHtmlProps> = props
             ry={computedCurve}
             style={{ transform: `translateX(${percent - 100}%)` }}
           />
+          {overAchieve && (
```
Comment (desfero) [inline app/components/shared/PercentageIndicatorBar.tsx]: if you merge with latest master this one is probably redundant, if not ping my on slack we should discuss this

File: app/modules/public-etos/reducer.ts (hunk 2 above)
Comment (Moejoe90) [reply app/modules/public-etos/reducer.ts]: How do you plan to simplify it 
@desfero ?

**Comment by rudolfix** [comment]
OK here is a heuristics we can use to eliminate most of the calls to nodes in `checkMaxCapExceeded`.
we need: 
1. maximum number of tokens sold: in whitelist it is  `new_shares_to_issue_in_whitelist`, in public phase `new_shares_to_issue`
2. token price without any discounts (which you already compute somewhere)
3. actual number of tokens sold (it's somewhere in eto progress state)

now if (maximum number of tokens sold - actual number of tokens sold) * token price < minimum ticket then we should call calculateContribution to get exact result

why it works:
max cap situation happens when so many tokens were sold that there's no way to buy more. as tokens can be bought with minimal amount of eur defined by min ticket we compute this as boundary value above

File: app/modules/investment-flow/sagas.ts [hunk 5]
```diff
@@ -186,7 +187,15 @@ function* validateAndCalculateInputs({ contractsService }: TGlobalDependencies):
     const etoContract: ETOCommitment = yield contractsService.getETOCommitmentContract(eto.etoId);
     if (etoContract) {
       const isICBM = selectIsICBMInvestment(state);
-      yield neuCall(loadComputedContributionFromContract, eto, value, isICBM);
+      const contribution = yield neuCall(loadComputedContributionFromContract, eto, value, isICBM);
+
+      yield put(
+        actions.investorEtoTicket.setCalculatedContribution(
+          eto.etoId,
+          convertToCalculatedContribution(contribution),
```
Comment (trival) [inline app/modules/investment-flow/sagas.ts]: please remove this call from here. here the data should already be transformed...

File: app/modules/investor-tickets/sagas.ts [hunk 6]
```diff
@@ -54,7 +54,14 @@ export function* loadInvestorTicket(
     ),
   );
 
-  yield neuCall(loadComputedContributionFromContract, action.payload.eto);
+  const contribution = yield neuCall(loadComputedContributionFromContract, action.payload.eto);
+
+  yield put(
+    actions.investorEtoTicket.setInitialCalculatedContribution(
+      etoId,
+      convertToCalculatedContribution(contribution),
```
Comment (trival) [inline app/modules/investor-tickets/sagas.ts]: also here

File: app/modules/investor-tickets/sagas.ts [hunk 7]
```diff
@@ -76,18 +83,11 @@ export function* loadComputedContributionFromContract(
 
     // TODO: check whether typechain but still is not fixed
     // sorry no typechain, typechain has a bug with boolean casting
-    const calculation = yield promisify(etoContract.rawWeb3Contract.calculateContribution, [
+    return yield promisify(etoContract.rawWeb3Contract.calculateContribution, [
```
Comment (trival) [inline app/modules/investor-tickets/sagas.ts]: I think this saga should already return the transformed data, not the raw array

File: app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
+import BigNumber from "bignumber.js";
+import * as cn from "classnames";
+import * as React from "react";
+import { FormattedMessage } from "react-intl-phraseapp";
+import { compose } from "redux";
```
Comment (desfero) [inline app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx]: move to `recompose`

File: app/components/eto/overview/EtoOverviewStatus/EtoOverviewStatus.tsx [hunk 9]
```diff
@@ -73,9 +78,26 @@ const EtoStatusManager = ({
...
+  maxCapExceeded,
 }: IExternalProps & IStateProps) => {
   // It's possible for contract to be undefined if eto is not on chain yet
   const timedState = eto.contract ? eto.contract.timedState : EETOStateOnChain.Setup;
+  const isEtoActive =
+    (isEligibleToPreEto && timedState === EETOStateOnChain.Whitelist) ||
+    timedState === EETOStateOnChain.Public;
+
+  if (maxCapExceeded && isEtoActive) {
+    if (!eto.contract) return null;
```
[the remark this answers is not in the document]
Comment (drakkein) [reply app/components/eto/overview/EtoOverviewStatus/EtoOverviewStatus.tsx]: Yeah, previously there was different condition, good catch

File: app/components/eto/overview/EtoOverviewStatus/EtoOverviewStatus.tsx [hunk 10]
```diff
@@ -73,9 +78,26 @@ const EtoStatusManager = ({
...
+
+    return (
+      <EtoMaxCapExceededWidget
+        eto={eto}
+        raisedTokens={eto.contract.totalInvestment.totalTokensInt.toNumber()}
```
[the remark this answers is not in the document]
Comment (drakkein) [reply app/components/eto/overview/EtoOverviewStatus/EtoOverviewStatus.tsx]: That would be easier

File: app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+export interface IExternalProps {
+  eto: TEtoWithCompanyAndContract;
+  raisedTokens: number;
+  totalEquivEurUlps: BigNumber;
+  totalInvestors: number;
+}
+
+interface IStateProps {
+  etherPriceEur: string;
```
[the remark this answers is not in the document]
Comment (drakkein) [reply app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx]: The reason was `any` props type in **EtoMaxCapExceededComponent**

File: app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  raisedTokens,
+  totalEquivEurUlps,
+  etherPriceEur,
+  totalInvestors,
+  isPreEto,
+}) => {
+  return (
+    <div className={styles.maxCapExceeded}>
+      <div className={cn(styles.header, styles.center)}>
+        <div>
```
[the remark this answers is not in the document]
Comment (drakkein) [reply app/components/eto/overview/EtoOverviewStatus/EtoMaxCapExceeded.tsx]: It's required for a medium size screen to behave correctly

**Pull request merged**